pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,227 @@
+package servicegraphprocessor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/grafana/agent/pkg/tempo/contextkeys"
+	"github.com/hashicorp/go-multierror"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/batchpersignal"
+	"github.com/patrickmn/go-cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	"go.uber.org/atomic"
+)
+
+var (
+	ErrNoServiceName = errors.New("failed to find service name")
+	ErrTooManyEdges  = errors.New("too many edges in memory")
+)
+
+type edge struct {
+	serverService, clientService string
+	serverLatency, clientLatency time.Duration
+
+	failed bool
+}
+
+func (e *edge) complete() bool {
+	return len(e.clientService) != 0 && len(e.serverService) != 0
+}
+
+var _ component.TracesProcessor = (*processor)(nil)
+
+type processor struct {
+	nextConsumer consumer.Traces
+	reg          prometheus.Registerer
+
+	store    *cache.Cache
+	maxEdges int
+
+	serviceGraphRequestTotal       *prometheus.CounterVec
+	serviceGraphRequestFailedTotal *prometheus.CounterVec
+	serviceGraphRequestHistogram   *prometheus.HistogramVec
+
+	closed atomic.Bool
+
+	logger log.Logger
+}
+
+func newProcessor(nextConsumer consumer.Traces, cfg *Config) (*processor, error) {
+	if cfg.wait == 0 {
+		cfg.wait = defaultWait
+	}
+	if cfg.maxEdges == 0 {
+		cfg.maxEdges = defaultMaxEdges
+	}
+
+	p := &processor{
+		nextConsumer: nextConsumer,
+		store:        cache.New(cfg.wait, cfg.wait*2),
+		maxEdges:     defaultMaxEdges,
+		closed:       atomic.Bool{},
+	}
+
+	return p, nil
+}
+
+func (p *processor) Start(ctx context.Context, _ component.Host) error {
+	var reg prometheus.Registerer
+	reg = prometheus.NewRegistry()
+	if v, ok := ctx.Value(contextkeys.PrometheusRegisterer).(prometheus.Registerer); ok {
+		reg = v
+	}
+	p.reg = reg
+
+	return p.registerMetrics()
+}
+
+func (p *processor) registerMetrics() error {
+	p.serviceGraphRequestTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_total",
+		Help: "Total count of requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestFailedTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_failed_total",
+		Help: "Total count of failed requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "tempo_service_graph_request_seconds",
+		Help:    "Time for a request between two nodes",
+		Buckets: prometheus.ExponentialBuckets(0.01, 2, 12),
+	}, []string{"client", "server"})
```

Comment:
an additional counter metric for unpaired edges would be nice. i.e. if i drop an edge due to the timeout or max edges before it gets paired.

we should also record the client/server on this metric as it might indicate an incoming or outgoing untraced edge.

Reply:
Added, thanks for the suggestion!